faiss/impl/AdditiveQuantizer.cpp

```diff
@@ -75,25 +96,94 @@ void AdditiveQuantizer::set_derived_values() {
         }
     }
     total_codebook_size = codebook_offsets[M];
+    switch (search_type) {
+        case ST_decompress:
+        case ST_LUT_nonorm:
+        case ST_norm_from_LUT:
+            break; // nothing to add
+        case ST_norm_float:
+            tot_bits += 32;
+            break;
+        case ST_norm_qint8:
+            tot_bits += 8;
+            break;
+        case ST_norm_qint4:
+            tot_bits += 4;
+            break;
+    }
+
     // convert bits to bytes
     code_size = (tot_bits + 7) / 8;
 }
 
+namespace {
+
+// TODO
+// https://stackoverflow.com/questions/31631224/hacks-for-clamping-integer-to-0-255-and-doubles-to-0-0-1-0
+
+uint8_t encode_qint8(float x, float amin, float amax) {
+    float x1 = (x - amin) / (amax - amin) * 256;
+    int32_t xi = int32_t(floor(x1));
+
+    return xi < 0 ? 0 : xi > 255 ? 255 : xi;
+}
+
+uint8_t encode_qint4(float x, float amin, float amax) {
+    float x1 = (x - amin) / (amax - amin) * 16;
+    int32_t xi = int32_t(floor(x1));
+
+    return xi < 0 ? 0 : xi > 15 ? 15 : xi;
+}
+
+float decode_qint8(uint8_t i, float amin, float amax) {
+    return (i + 0.5) / 256 * (amax - amin) + amin;
+}
+
+float decode_qint4(uint8_t i, float amin, float amax) {
+    return (i + 0.5) / 16 * (amax - amin) + amin;
+}
```

Comment:
We could replace these with a non-uniform quantization using 1-D K-means. But I think it would not improve the performance too much for 8 bits code. I can do it in a follow-up PR.